src/app/lecture/knowledge-node/notes/markdown-editor/markdown-editor.component.ts

```diff
@@ -0,0 +1,137 @@
+import { Component, Input, ViewChild, forwardRef } from '@angular/core';
+import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
+
+@Component({
+  selector: 'cc-markdown-editor',
+  templateUrl: './markdown-editor.component.html',
+  styleUrls: ['./markdown-editor.component.css'],
+  providers: [
+    {
+      provide: NG_VALUE_ACCESSOR,
+      useExisting: forwardRef(() => MarkdownEditorComponent),
+      multi: true
+    }
+  ]
+})
+export class MarkdownEditorComponent implements ControlValueAccessor {
+
+  text = '';
+  livePreview = true;
+  selection: any;
+  @ViewChild('textAreaElement') textArea: HTMLTextAreaElement;
+  @Input() mode = 'edit';
+
+  constructor() { }
+
+  get value(): string {
+    return this.text;
+  }
+
+  set value(value: string) {
+    this.text = value;
+    this.updateChanges();
+  }
+
+  updateChanges(): void {
+    this.onChange(this.value);
+    this.onTouched();
+  }
+
+  onChange: any = () => {};
+
+  onTouched: any = () => {};
+
+  writeValue(obj: any): void {
+    if (obj) {
+      this.value = obj;
+    } else {
+      this.value = '';
+    }
+  }
+
+  registerOnChange(fn: any): void {
+    this.onChange = fn;
+  }
+
+  registerOnTouched(fn: any): void {
+    this.onTouched = fn;
+  }
+
+  setDisabledState?(isDisabled: boolean): void {
+    this.textArea.disabled = isDisabled;
+  }
+
+  insertElement(type: string): void {
+    let tagBegin = '';
+    let tagEnd = '';
+    let tagText = '';
+    switch (type) {
+      case 'bold':
+        tagBegin = tagEnd = '**';
+        tagText = 'Bold Text';
+        break;
+      case 'italic':
+        tagBegin = tagEnd = '_';
+        tagText = 'Italic Text';
+        break;
+      case 'bulleted':
+        tagBegin = '- ';
+        tagText = 'List Item';
+        break;
+      case 'numbered':
+        tagBegin = '1. ';
+        tagText = 'List Item';
+        break;
+      case 'code':
+        tagBegin = tagEnd = '```';
+        tagText = 'Code';
+        break;
+      case 'h1':
+        tagBegin = '# ';
+        tagText = 'Heading 1';
+        break;
+      case 'h2':
+        tagBegin = '## ';
+        tagText = 'Heading 2';
+        break;
+      case 'h3':
+        tagBegin = '### ';
+        tagText = 'Heading 3';
+        break;
+      case 'h4':
+        tagBegin = '#### ';
+        tagText = 'Heading 4';
+        break;
+      case 'h5':
+        tagBegin = '##### ';
+        tagText = 'Heading 5';
+        break;
+      case 'h6':
+        tagBegin = '###### ';
+        tagText = 'Heading 6';
```

Comment:
Considering the use of the widget, I think h1 and h2 are enough, but we can leave h3 just in case. The rest are too much.